web/profiles/custom/yalesites_profile/modules/custom/ys_templated_content/src/Importers/ZipFileImporter.php

```diff
@@ -0,0 +1,245 @@
+          '\Drupal\single_content_sync\ContentBatchImporter::batchImportAssets',
+          [$original_file_path, $zip_file],
+        ];
+      }
+    }
+
+    foreach ($zip->listContents() as $zip_file) {
+      $original_file_path = "{$import_directory}/{$zip_file}";
+
+      if ($this->isZipFileValid($original_file_path)) {
+        $content_file_path = $original_file_path;
+        $batch['operations'][] = [
+          '\Drupal\single_content_sync\ContentBatchImporter::batchImportFile',
+          [$original_file_path],
+        ];
+      }
+    }
+
+    if (!$batch['operations']) {
+      throw new \Exception(' Please check the structure of the zip file and ensure you do not have an extra parent directory.');
+    }
+
+    $batch['operations'][] = [
+      '\Drupal\single_content_sync\ContentBatchImporter::cleanImportDirectory',
+      [$import_directory],
+    ];
+
+    $batch['operations'][] = [
+      '\Drupal\single_content_sync\ContentBatchImporter::cleanImportDirectory',
+      [$tempDir],
+    ];
+
+    if (is_null($content_file_path)) {
+      throw new \Exception('The content file in YAML format could not be found.');
+    }
+
+    batch_set($batch);
+
+    return NULL;
+  }
+
+  /**
+   * Validate zip file before we run batch.
+   *
+   * Taken from Single Content Sync module: @see https://git.drupalcode.org/project/single_content_sync/-/blob/1.4.x/src/ContentImporter.php?ref_type=heads#L348-371
+   *
+   * @param string $path
+   *   The local file path of the extracted zip file.
+   *
+   * @return bool
+   *   TRUE if the valid YML file is found.
+   */
+  protected function isZipFileValid(string $path): bool {
+    $info = pathinfo($path);
+
+    if (is_file($path) && $info['extension'] === 'yml') {
+      // Extra directory found, let's skip the operation and trigger
+      // an error later.
+      [, $directory] = explode('://', $info['dirname']);
+
+      // If there are more than 3 parts, then there is an extra folder.
+      // e.g. import/zip/uuid is correct one.
+      if (count(explode('/', $directory)) > 3) {
+        return FALSE;
+      }
+
+      // File name can't start with dot.
+      if (strpos($info['filename'], '.') === 0) {
+        return FALSE;
+      }
+
+      return TRUE;
+    }
+
+    return FALSE;
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * Redirect to the entity created.
+   */
+  public static function importFinished($success, $results, $operations) {
+    if ($success) {
+      // Get the last node created to redirect to.
+      $query = \Drupal::entityQuery('node')
+        ->sort('changed', 'DESC')
+        ->range(0, 1)
+        ->accessCheck(FALSE);
+      $nids = $query->execute();
+      $nid = reset($nids);
+
+      return new RedirectResponse(
+      Url::fromRoute(
+      'entity.node.edit_form',
+      ['node' => $nid],
+      )->toString()
+      );
+    }
+    else {
+      drupal_set_message(t('An error occurred and the content could not be imported.'), 'error');
```

Comment:
We should replace this with the Messenger service. Not sure when this function is going away, but I feel like it's been on the chopping block for a long time but never actually gets depricated.

Reply:
Can do; I think I originally used it as there was some issue that I can't remember with it being inside of the batch job queue.   But I'll figure it out--sounds like we'd have to make the change anyway in the future, so might as well get it now.

Is it that way with a lot of those druapl_<function> methods or is this one special?  If it's most all of them at this point I'll try to steer clear of them all in the future.